pallets/dapp-staking/src/lib.rs

```diff
@@ -2707,55 +2707,60 @@ pub mod pallet {
         /// If no progress is made, `Err(_)` is returned.
         pub fn do_update(weight_limit: Weight) -> Result<Weight, Weight> {
             // Find out if update is still in progress
-            let maybe_cursor = ActiveBonusUpdateCursor::<T>::get();
+            let init_bonus_update_state = ActiveBonusUpdateState::<T>::get();
             let mut consumed_weight = T::DbWeight::get().reads(1);
-            let mut new_cursor = None;
+
+            if init_bonus_update_state == BonusUpdateStateFor::<T>::Finished {
+                log::trace!(
+                    target: LOG_TARGET,
+                    "Update has finished, skipping any action."
+                );
+                return Err(consumed_weight);
+            }
+
+            let mut bonus_update_state = init_bonus_update_state.clone();
             let mut entries_updated = 0u32;
 
-            let mut iter = if let Some(last_key_pair) = maybe_cursor.clone() {
-                StakerInfo::<T>::iter_from(StakerInfo::<T>::hashed_key_for(
-                    last_key_pair.0,
-                    last_key_pair.1,
-                ))
-            } else {
-                // Enable maintenance mode on the first run
-                // Likely to be already set in the runtime upgrade migration
-                ActiveProtocolState::<T>::mutate(|state| {
-                    state.maintenance = true;
-                });
-                log::warn!("Maintenance mode enabled.");
-                consumed_weight.saturating_add(T::DbWeight::get().reads_writes(1, 2));
-                StakerInfo::<T>::iter()
-            };
+            let mut iter =
+                if let BonusUpdateStateFor::<T>::InProgress(last_key_pair) = &bonus_update_state {
```

Comment:
`if let BonusUpdateStateFor::<T>::InProgress((account, contract)) = &bonus_update_state {`
can you make it more explicit?

Reply:
I've destructured the tuple, it should be more explicit now